pkg/node/manager/manager.go

```diff
@@ -226,11 +241,105 @@ func New(c Configuration, ipCache IPCache, nodeMetrics *nodeMetrics, healthRepor
 	return m, nil
 }
 
-func (m *manager) Start(hive.HookContext) error {
+func (m *manager) Start(ctx hive.HookContext) error {
+	if err := m.restoreExistingIdentities(ctx); err != nil {
+		log.WithError(err).Warn("failed to restore existing node CIDR identities")
+	}
 	m.workerpool = workerpool.New(numBackgroundWorkers)
 	return m.workerpool.Submit("backgroundSync", m.backgroundSync)
 }
 
+// restoreExistingIdentities takes any already-existing identities in the BPF ipcache and
+// back-fills them in to the userspace ipcache.
+// This is necessary to prevent policy "blips" on startup by preventing accidental re-use of
+// an identity still in the bpf ipcache and policymaps. It roughly matches the existing
+// CIDR identity restoration flow.
+func (m *manager) restoreExistingIdentities(ctx context.Context) error {
+
+	// Do not restore node identities if they are not selectable by CIDR policy
+	if !option.Config.PolicyCIDRSelectsNodes {
+		return nil
+	}
+
+	// Dump existing ip-identity pairs from the ipcache
+	existingPrefixes := map[netip.Prefix]identity.NumericIdentity{}
+	err := ipcachemap.IPCacheMap().DumpWithCallback(func(key bpf.MapKey, value bpf.MapValue) {
```

Comment:
Dumping a large map another time during startup seems like unnecessary work. I see at the top of the function that there's an option config check to protect this and try to isolate this logic, but I think there are other ways to achieve this without adding all of this logic to node manager. I'm also not entirely sure why the node manager needs to deal with so many deep details of the datapath. Depending directly on the ipcachemap seems like a layering violation.

Reply:
This has been done, resolving this comment.